### DISCUSSION TITLE
Totale errato su Fatture in Cloud

### DISCUSSION THREAD

**Discussion by Drakas** [fattureincloud/api#109] [Q&A]
Buongiorno,
Sto collegando Prestashop utilizzando il plugin officiale per creare le fatture su Fatture in Cloud.
Noto il seguente problema di arrotondamento quando si applica lo sconto.

Ordine creato su Prestahop:
1xPRODOTTO TEST FATTURE, prezzo netto: 975,409836 con poi applicazione del 50% di sconto

Considerando l'IVA 22%, il totale finito lordo per prestashop è **595,00 €** , valore corretto e questo è quello che si vorrebbe anche come valore finale sulla fattura di Fatture in Cloud, invece viene generata la Fattura con lordo uguale a **594,99 €** . Il che è un problema dato che c'è quel centesimo.

Come è possibile risolvere per avere tale valore finale corretto anche su Fatture in Cloud? (ovvero **595,00 €** )

In calce allego anche i dati trasmessi e ricevuti via API: Plugin Prestashop <---> Fatture in Cloud.

Segnalo che, se entro in Fatture in Cloud, poi dentro la singola fattura e spunto "Utilizza prezzi lordi", allora il totale è corretto. Naturalmente però non posso farlo per ogni fattura, altrimenti perderebbe di senso l'automazione.
Allego screenshot con fattura aperta e fattura aperta con spunta "Utilizza prezzi lordi".

<img width="884" alt="Immagine 2022-10-06 193937" src="https://user-images.githubusercontent.com/3465823/194381958-8d4a1bf9-d11b-4172-9bbc-a762dbf17e3e.png">
<img width="853" alt="Immagine 2022-10-06 193947" src="https://user-images.githubusercontent.com/3465823/194381953-553ea0e7-1e88-4318-8125-90c36817196a.png">

Esempio Array usato come parametro per creare la Fattura su Fatture in Cloud
`{
  "data": {
    "currency": {
      "id": "EUR",
      "exchange_rate": "1.00000"
    },
    "entity": {
      "name": "test",
      "address_street": "sadasd",
      "address_postal_code": "22222",
      "address_city": "werwer",
      "address_province": "AN",
      "country_iso": "IT",
      "shipping_address": "",
      "tax_code": "234234324",
      "code": "PS-34 - 166",
      "mail": "test@test.com",
      "phone": "324324234",
      "id": "62217752",
      "certified_email": "",
      "ei_code": ""
    },
    "items_list": [
      {
        "code": "TESTING-FATTURAZIONE",
        "name": "PRODOTTO TEST FATTURE",
        "qty": "1",
        "net_price": "975.409836",
        "discount": "50.00",
        "vat": {
          "id": 0
        }
      }
    ],
    "visible_subject": "",
    "subject": "Ordine ###",
    "payments_list": [
      {
        "due_date": "2022-10-01 14:14:14",
        "status": "paid",
        "paid_date": "0000-00-00 00:00:00",
        "payment_account": {
          "id": 12345
        }
      }
    ],
    "payment_method": {
      "name": "Bonifico bancario"
    },
    "show_payment": false,
    "show_payment_method": true,
    "rc_center": "",
    "notes": "###",
    "e_invoice": true,
    "ei_data": {
      "payment_method": "MP05"
    },
    "type": "invoice",
    "numeration": "TESTING"
  },
  "options": {
    "fix_payments": true
  }
}`

E di seguito esempio di risposta da Fatture in Cloud data la fattura creata con i dati sopra (ID e altri dati non utili, sono anonimizzati):

`{
  "data": {
    "id": 123123,
    "type": "invoice",
    "year": 2022,
    "numeration": "TESTING",
    "subject": "Ordine #",
    "visible_subject": "",
    "rc_center": "",
    "amount_rivalsa": 0,
    "amount_rivalsa_taxable": 487.7,
    "amount_global_cassa_taxable": 487.7,
    "amount_cassa": 0,
    "amount_cassa_taxable": 487.7,
    "amount_cassa2": 0,
    "amount_cassa2_taxable": 487.7,
    "amount_withholding_tax": 0,
    "amount_withholding_tax_taxable": 487.7,
    "amount_other_withholding_tax": 0,
    "amount_enasarco_taxable": 0,
    "amount_other_withholding_tax_taxable": 487.7,
    "ei_cassa_type": null,
    "ei_cassa2_type": null,
    "ei_withholding_tax_causal": null,
    "ei_other_withholding_tax_type": null,
    "ei_other_withholding_tax_causal": null,
    "stamp_duty": 0,
    "use_gross_prices": false,
    "e_invoice": true,
    "agyo_company_id": null,
    "agyo_id": null,
    "agyo_sent_at": null,
    "delivery_note": false,
    "accompanying_invoice": false,
    "amount_net": 487.7,
    "amount_vat": 107.29,
    "amount_gross": 594.99,
    "amount_due_discount": 0,
    "permanent_token": "123123",
    "h_margins": 15,
    "v_margins": 16,
    "show_payment_method": true,
    "show_payments": true,
    "show_totals": "all",
    "show_paypal_button": true,
    "show_notification_button": false,
    "is_marked": false,
    "created_at": "2022-10-06 17:04:43",
    "updated_at": "2022-10-06 17:04:43",
    "entity": {
      "id": 123123,
      "name": "123123",
      "vat_number": "",
      "tax_code": "234234324",
      "address_street": "sadasd",
      "address_postal_code": "22222",
      "address_city": "werwer",
      "address_province": "AN",
      "address_extra": "",
      "country": "Italia",
      "certified_email": "",
      "ei_code": "0000000",
      "type": null
    },
    "date": "2022-10-06",
    "number": 1,
    "currency": {
      "id": "EUR",
      "exchange_rate": "1.00000",
      "symbol": "€"
    },
    "language": {
      "code": "it",
      "name": "Italiano"
    },
    "notes": "123123",
    "rivalsa": 0,
    "rivalsa_taxable": 100,
    "global_cassa_taxable": 100,
    "cassa": 0,
    "cassa_taxable": 100,
    "cassa2": 0,
    "cassa2_taxable": 100,
    "withholding_tax": 0,
    "withholding_tax_taxable": 100,
    "other_withholding_tax": 0,
    "other_withholding_tax_taxable": 100,
    "payment_method": {
      "id": null,
      "name": "Bonifico bancario",
      "details": [
        {
          "title": "",
          "description": ""
        },
        {
          "title": "",
          "description": ""
        },
        {
          "title": "",
          "description": ""
        },
        {
          "title": "",
          "description": ""
        },
        {
          "title": "",
          "description": ""
        }
      ]
    },
    "use_split_payment": false,
    "merged_in": null,
    "original_document": null,
    "ei_data": {
      "vat_kind": null,
      "original_document_type": null,
      "od_number": "",
      "od_date": null,
      "cig": "",
      "cup": "",
      "payment_method": "MP05",
      "bank_name": "",
      "bank_iban": "",
      "bank_beneficiary": ""
    },
    "ei_ts_data": {
      "status": 0,
      "id": null,
      "info": null
    },
    "ei_raw": null,
    "ei_status": "not_sent",
    "items_list": [
      {
        "product_id": null,
        "code": "TESTING-FATTURAZIONE",
        "name": "PRODOTTO TEST FATTURE",
        "measure": "",
        "net_price": 975.40984,
        "category": "",
        "id": 123123,
        "gross_price": 1190,
        "apply_withholding_taxes": false,
        "discount": 50,
        "discount_highlight": false,
        "in_dn": false,
        "qty": 1,
        "vat": {
          "id": 0,
          "value": 22,
          "description": ""
        },
        "stock": true,
        "description": "",
        "not_taxable": false,
        "ei_raw": null
      }
    ],
    "payments_list": [
      {
        "amount": 594.99,
        "due_date": "2022-10-01",
        "paid_date": "0000-00-00",
        "id": 123123,
        "forfettari_revenue": 594.99,
        "payment_terms": {
          "days": 0,
          "type": "standard"
        },
        "status": "paid",
        "payment_account": {
          "id": 123123,
          "name": "Bonifico bancario",
          "virtual": false
        },
        "ei_raw": null
      }
    ],
    "attachment_url": null,
    "seen_date": null,
    "next_due_date": null,
    "template": {
      "id": 123123,
      "name": "13123"
    },
    "extra_data": null,
    "url": "https://compute.fattureincloud.it/doc/123123.pdf",
    "locked": false,
    "ei_locked": false,
    "has_ts_pay_pending_payment": false,
    "show_tspay_button": false,
    "pay_with_tspay_url": null
  }
}`

**Comment by emajo** [comment]
Ciao @Drakas abbiamo girato la tua richiesta allo sviluppatore dell'integrazione con prestashop, le faremo sapere appena abbiamo aggiornamenti.

Grazie per la pazienza,
Emanuele

**Comment by Drakas** [comment]
Ciao @emajo , grazie mille, segnalo però che questo errore si verifica anche se si utilizzano comunque le Vostre API direttamente (senza plugin).
(ovvero, l'impossibilità di ottenere il lordo corretto di 595 euro se viene creata la fattura con netto pari a 975,409836 e sconto 50%). 

A presto

**Reply by valmoz** [reply]
Buongiorni @Drakas 
ha letto la [guida dedicata ai totali](https://developers.fattureincloud.it/docs/guides/invoice-totals)?

Grazie
Mauro

**Reply by Drakas** [reply]
Buongiorno @valmoz , si l'avevo già letta, però purtroppo non c'è molto che possa risolvere nel nostro caso.

Ti riferivi a qualcosa in particolare?

Grazie

**Reply by emajo** [reply]
Ciao @Drakas, purtroppo è questione di arrotondamenti su cui non possiamo andare ad agire, l'unico modo per risolvere sarebbe implementare nell'integrazione l'utilizzo dei prezzi lordi.

**Reply by Drakas** [reply]
Ciao @emajo , per risolvere in modo veloce, è possibile simulare quel "Utilizza prezzi lordi" che c'è nel frontend (vedi immagini allegate all'inizio) e che risolve ma via API? Perché basterebbe quello per risolvere il problema. 

Esempio:
1) creo la fattura via API con corrispettivo ID
2) Tramite tale ID, inoltro la richiesta di modifica del documento con "Utilizza prezzi lordi"

Ho già provato a farlo, tramite il parametro use_gross_prices = true , ma mi va in errore dicendo che i conti non tornano (immagino perché non avvengono gli automatismi che invece vi sono nel frontend al click del flag). Si può fare qualcosa in merito? Sarebbe estremamente risolutivo e utile.

Grazie mille e  a presto

**Reply by emajo** [reply]
Ciao @Drakas, putroppo come dici tu quegli automatismi sono proprio fatti dal nostro frontend e non c'è modo per replicarli tramite api.

**Reply by Drakas** [reply]
Ciao @emajo , grazie mille per la conferma, immaginavo. 
Per caso c'è qualche risposta dallo sviluppatore dell'integrazione con prestashop? dato che tale plugin lavora con i netti, certamente è da riprogrammare con il lordo per ottenere il risultato finale corretto. Apro anche la issue su tale repository
